modules/stochastic_tools/src/state_sim/model/actions/StateSimActTransition.C

```diff
@@ -0,0 +1,167 @@
+    _to_state_probs(),
+    _to_state_var_probs(),
+    _fail_desc(),
+    _mut_excl(false),
+    _has_var_probs(false)
+{
+  this->addToState(to_state, -1, move_desc);
+}
+
+void
+StateSimActTransition::addToState(StateSimState & to_state, StateSimBase * state_sim_var_prob, const std::string & fail_desc)
+{
+  if ((state_sim_var_prob == NULL) ||
+      (typeid(*state_sim_var_prob) != typeid(StateSimVariable)) ||
+      (((StateSimVariable *)state_sim_var_prob)->getVarType() != VT_DOUBLE))
+  {
+    mooseAssert(false, "state_sim_var_prob parameter must be a double *");
+    //throw std::invalid_argument("state_sim_var_prob parameter must be a double *");
+  }
+
+  _has_var_probs = true;
+  //StateSimVariable<double*> *prob_var = dynamic_cast<StateSimVariable<double*> *>(state_sim_var_prob);
+  StateSimVariable * prob_var = (StateSimVariable *)state_sim_var_prob;
+
+  _to_state_probs.push_back(prob_var->getDouble());
+  _to_state_var_probs.push_back(prob_var);
+  _to_states.push_back(&to_state);
+  _fail_desc.push_back(fail_desc);
+}
+
+void
+StateSimActTransition::addToState(StateSimState & to_state, const double & prob, const std::string & fail_desc)
+{
+  if (prob < 0) //use the remainder so put it on the back of the lise.
+  {
+    _to_state_probs.push_back(prob);
+    _to_state_var_probs.push_back(NULL);
+    _to_states.push_back(&to_state);
+    _fail_desc.push_back(fail_desc);
+  }
+  else
+  {
+    _to_state_probs.insert(_to_state_probs.begin(), prob);
+    _to_state_var_probs.insert(_to_state_var_probs.begin(), NULL);
+    _to_states.insert(_to_states.begin(), &to_state);
+    _fail_desc.insert(_fail_desc.begin(), fail_desc);
+  }
+}
+
+void
+StateSimActTransition::removeToState(StateSimState & rem_state)
+{
+  StateSimState * a = &rem_state;
+  //auto func = [&a](const StateSimState * b) -> bool {return a==b;};
+  auto found_state = std::find_if(_to_states.begin(), _to_states.end(), [&a](const StateSimState * b) { return a == b; });
+  if (found_state != _to_states.end())
+  {
+    int idx = found_state - _to_states.begin();
+    _to_states.erase(found_state);
+    _to_state_probs.erase(_to_state_probs.begin() + idx);
+    _to_state_var_probs.erase(_to_state_var_probs.begin() + idx);
+    _fail_desc.erase(_fail_desc.begin() + idx);
+  }
+}
+
+std::vector<std::shared_ptr<StateSimIdAndDesc>>
+StateSimActTransition::WhichToState()
+{
+  std::vector<std::shared_ptr<StateSimIdAndDesc>> ret_list;
+
+  if (_has_var_probs)
+  {
+    for (unsigned int i = 0; i < _to_state_var_probs.size(); i++)
+    {
+      if (_to_state_var_probs[i] != NULL)
+      {
+        _to_state_probs[i] = ((StateSimVariable *)_to_state_var_probs[i])->getDouble();
+      }
+    }
+  }
+
+  double prob_sum = 0;
+  for (auto & n : _to_state_probs)
+  {
+    prob_sum += n;
+  }
+
+  if (_mut_excl && (prob_sum > 0) && (prob_sum < 1.0))
+    mooseAssert(false, "No default state entered for " + this->name() + " and transitions probabilites don't add up to 1.0.");
```

Comment:
`mooseAssert` gets completely removed in optimized mode. We are transitioning to a new system but ONLY having a mooseAssert in an `if` statement is a bad idea (I'm surprised it even compiles). If you have code that you only need to run in debug mode you can use `#ifndef NDEBUG` or `#ifdef DEBUG` depending on the kind of check and how expensive it is. See me for more details on this.